src/predictor/gpu_predictor.cu

```diff
@@ -262,31 +222,81 @@ class GPUPredictor : public xgboost::Predictor {
   };
 
  private:
-  void DevicePredictInternal(DMatrix* dmat,
-                             HostDeviceVector<bst_float>* out_preds,
-                             const gbm::GBTreeModel& model, size_t tree_begin,
-                             size_t tree_end) {
-    if (tree_end - tree_begin == 0) {
-      return;
+  void DeviceOffsets(const HostDeviceVector<size_t>& data, std::vector<size_t>* out_offsets) {
+    auto& offsets = *out_offsets;
+    offsets.resize(devices.Size() + 1);
+    offsets[0] = 0;
+#pragma omp parallel for schedule(static, 1) if (devices.Size() > 1)
```

Comment:
Sorry for dropping by.  Just to be safe, it might be better to save the current device before spawning threads that can change it.  Otherwise subsequent code could potentially access memory at wrong device.

https://github.com/dmlc/xgboost/blob/baef5741dfeec121dbeca9802e6b00a75ed72d03/src/common/device_helpers.cuh#L947

Reply:
`cudaSetDevice()` changes the device of the caller thread only, so it does not alter the device in a different thread.

In case only a single device is used (and no threads are spawned), `cudaSetDevice()` will be called with that device.

Otherwise, the code using a GPU is responsible for setting the device being used. This means `cudaSetDevice()` in public methods, in shards and in OpenMP loops (with 1 iteration per device). Private methods can assume that the right device has been set already.